src/main/java/com/google/sps/data/algorithm/TspSolver.java

```diff
@@ -0,0 +1,414 @@
+   * 
+   * @param center central location
+   * @param pois list of pois other than central location
+   */
+  private void createTimeMatrix(String center, List<String> pois) 
+      throws IOException {
+
+    // create allPois to hold `center` and all other `pois`
+    List<String> allPois = new ArrayList<>();
+    allPois.add(center);
+    allPois.addAll(pois);
+
+    // create timeMatrix and fill diagonal with 0's
+    timeMatrix = new int[allPois.size()][allPois.size()];
+    for (int i = 0; i < allPois.size(); i++) {
+      timeMatrix[i][i] = 0;
+    }
+
+    // fill (i, j) and (j, i) indices with travel time from i to j
+    for (int i = 0; i < allPois.size(); i++) {
+      for (int j = i; j < allPois.size(); j++) {
+        int travelTime = getTravelTimeMins(allPois.get(i), allPois.get(j));
+        timeMatrix[i][j] = travelTime;
+        timeMatrix[j][i] = travelTime;
+      }
+    }
+  }
+
+  /**
+   * Populate placeIdToInt and call setOpenHours for each placeId
+   *
+   * @param center central location
+   * @param pois list of pois to visit
+   */ 
+  private void populateIntMapAndOpenHours(String center, 
+      List<String> pois) throws IOException{
+    this.openHours = new HashMap<>();
+    this.intToPlaceId = new HashMap<>();
+    this.intToPlaceId.put(0, center);
+    setOpenHours(0, center);
+
+    for (int i = 0; i < pois.size(); i++) {
+      this.intToPlaceId.put(i+1, pois.get(i));
+      setOpenHours(i+1, pois.get(i));
+    }
+  }
+
+  /**
+   * Sets the openHours for each placeId (as int), as an OpeningHours.Period
+   * If can't find opening hours set to be null (represents open all day)
+   * 
+   * @param placeId placeId to search openHours for
+   * @param index index to put the openHours in the HashMap
+   */
+  private void setOpenHours(int index, String placeId) throws IOException {
+    PlaceDetails placeDetails = TripServlet.getPlaceDetailsFromPlaceId(this.context, placeId);
+
+    if (placeDetails.openingHours == null) { 
+      this.openHours.put(index, null);    
+    } else if (placeDetails.openingHours.periods.length != 7) {
+      checkDayIsPresentInHours(placeDetails.openingHours, index);
+    } else {
+      this.openHours.put(index, placeDetails.openingHours.periods[this.intOfWeek]);
+    } 
+  }
+
+  /**
+   * If day in openingHours array is the same as the day searched for 
+   * check if open and close time are not null. If no hours found put null
+   * into openHours (represents open all day)
+   * 
+   * @param openingHours openHours for placeId on desired day
+   * @param index index to put result in OpenHours Map
+   */
+  private void checkDayIsPresentInHours(OpeningHours openingHours, int index) {
+      int adjustedWeekIndex;
+      // periods is numbered Sunday - Saturday, WeekdayText is Monday - Sunday
+      if (intOfWeek == 0) {
+        adjustedWeekIndex = 6;
+      } else {
+        adjustedWeekIndex = intOfWeek - 1;
+      }
```

Comment:
totally optional, but check out the modulus operator (`%`) for an alternate way to do this. Eg: `(intOfWeek + 6) % 7`

Reply:
I should have thought about using mod that's much cleaner!